### PR TITLE
source-application: add recording support when streaming quality drops

### PR DESCRIPTION
**Pull request by raghaven447**
The logic to start/stop the recording on streaming quality drop/regain has been added into the agent (Gaeul).

### CODE REVIEW

**Comment by raghaven447** [comment]
I believe that the support to set the adaptor type on the GaeguliPipeline according to user choice should be added here in order to enable/disable the recording when streaming quality dropd/regains. Please comment.

File: gaeul/source/source-application.c [hunk 1]
```diff
@@ -41,8 +42,11 @@ typedef struct _GaeguliNest
   gint refcount;
 
   gchar *id;
+  gchar *record_location;
```
Comment (justinjoy) [inline gaeul/source/source-application.c:45]: This pointer should be freed in `gaeguli_nest_unref()`.

File: gaeul/source/source-application.c [hunk 2]
```diff
@@ -250,6 +254,42 @@ gaeguli_nest_dbus_register (GaeguliNest * nest, GDBusConnection * connection)
...
+}
+
+static void
+_stream_quality_regained_cb (GaeguliNest * nest)
+{
+  g_autoptr (GError) error = NULL;
+
+  gaeguli_pipeline_remove_target (nest->pipeline, nest->record_stream, &error);
+
+  g_debug ("target removed");
```
Comment (justinjoy) [inline gaeul/source/source-application.c]: `error->message` may have valuable information for debuggers.

File: gaeul/source/source-application.c [hunk 3]
```diff
@@ -250,6 +254,42 @@ gaeguli_nest_dbus_register (GaeguliNest * nest, GDBusConnection * connection)
...
 
+static void
+_stream_quality_dropped_cb (GaeguliNest * nest)
+{
+  g_autoptr (GError) error = NULL;
+
+  nest->record_stream =
+      gaeguli_pipeline_add_recording_target (nest->pipeline,
+      nest->record_location, &error);
+  gaeguli_target_start (nest->record_stream, &error);
```
Comment (justinjoy) [inline gaeul/source/source-application.c:269]: Can we reuse `error` without explicit free?

File: gaeul/source/source-application.c [hunk 4]
```diff
@@ -250,6 +254,42 @@ gaeguli_nest_dbus_register (GaeguliNest * nest, GDBusConnection * connection)
...
 }
 
+static void
+_stream_quality_dropped_cb (GaeguliNest * nest)
+{
+  g_autoptr (GError) error = NULL;
+
+  nest->record_stream =
+      gaeguli_pipeline_add_recording_target (nest->pipeline,
+      nest->record_location, &error);
```
Comment (justinjoy) [inline gaeul/source/source-application.c:264]: Please add error handling. `gaeguli_pipeline_add_recording_target()` can return NULL.

File: gaeul/source/source-application.c [hunk 5]
```diff
@@ -41,8 +42,11 @@ typedef struct _GaeguliNest
   gint refcount;
 
   gchar *id;
+  gchar *record_location;
   GaeguliPipeline *pipeline;
   GaeguliTarget *target_stream;
+  GaeguliTarget *record_stream;
```
Comment (justinjoy) [inline gaeul/source/source-application.c:48]: This should be freed when stopping.

File: gaeul/source/source-application.c [hunk 6]
```diff
@@ -41,8 +42,11 @@ typedef struct _GaeguliNest
...
+  GaeguliStreamAdaptor *adaptor;
```
Comment (justinjoy) [inline gaeul/source/source-application.c:49]: This should be freed in `gaeguli_nest_unref()`.

File: gaeul/source/source-application.c [hunk 7]
```diff
@@ -250,6 +254,42 @@ gaeguli_nest_dbus_register (GaeguliNest * nest, GDBusConnection * connection)
...
+  g_debug ("target removed");
+}
+
+static void
+_stream_started_cb (GaeguliPipeline * pipeline, guint target_id,
+    gpointer user_data)
+{
+  GaeguliNest *nest = user_data;
+
+  nest->adaptor = gaeguli_target_get_stream_adaptor (nest->target_stream);
```
Comment (justinjoy) [inline gaeul/source/source-application.c:300]: I am not sure whether the pipeline emits stream-start only once.
We should avoid memory leakage.

**Review by justinjoy — CHANGES_REQUESTED** [review]

**Comment by justinjoy** [comment]
> I believe that the support to set the adaptor type on the GaeguliPipeline according to user choice should be added here in order to enable/disable the recording when streaming quality dropd/regains. Please comment.

Yes, the recording feature can be enabled/disabled by user, but since we have only a bandwidth adaptor,  I don't think we need to provide an option to choose a different adaptor now.

**Review by justinjoy — APPROVED** [review]
LGTM

**Pull request merged**